### ISSUE TITLE
Calling `torch.tensor.item()` renders execution very slow

### ISSUE THREAD

**Issue by patrickvonplaten** [pytorch/xla#2120]
## 🐛 Bug

I am trying to train a `transformers` model in a google colab on TPU. 
When running all operations as tensors the execution time seems reasonable. 

As soon as I call `torch.tensor.item()` at the end of the script it becomes ~100 times slower.

## To Reproduce

1. I install the nightly version in a google colab via

```python 
import os
assert os.environ['COLAB_TPU_ADDR'], 'Make sure to select TPU from Edit > Notebook settings > Hardware accelerator'

VERSION = "nightly"  #@param ["1.5" , "20200325", "nightly"]
!curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
!python pytorch-xla-env-setup.py --version $VERSION
```

2. I install `transformers`:
```python
# install transformers
!pip install git+https://github.com/huggingface/transformers.git -qq
```

3. I execute the following code:
```python 
from transformers import ReformerModelWithLMHead, ReformerConfig
import os
import time
import torch

import torch_xla
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met

os.environ["XLA_USE_32BIT_LONG"] = "1"
device = xm.xla_device()

length_pow = 7
epochs = 5

# load config and model
config = ReformerConfig(axial_pos_shape=[2**(length_pow-3), 2**3], max_position_embeddings=2**length_pow)

# initialize model
model = ReformerModelWithLMHead(config)
model.train()
model.to(device)

# start timer
start_time = time.time()

sum_loss = 0
for step in range(epochs):
  input_ids = torch.tensor([(2 ** length_pow) * [0]], device=device, dtype=torch.long)
  loss = model(input_ids, labels=input_ids)[0]
  loss.backward()
  sum_loss += loss

avg_loss = sum_loss / epochs

# UNCOMMENTING THIS LINE RENDERS THE NOTEBOOK SUPER SLOW - WHY?
# print("Avg Loss: {}".format(avg_loss.item()))

# end timer
end_time = time.time() - start_time
print("Time: {}".format(end_time))

print(met.metrics_report())
```

When commenting out the line `print("Avg Loss: {}".format(avg_loss.item()))` the notebook executes in 0.3s which is reasonable. When running the notebook and printing the average loss (uncommenting the line: `# print("Avg Loss: {}".format(avg_loss.item()))`) the notebook executes in ~30 seconds.

The problem can be reproduced easily when running this colab: https://colab.research.google.com/drive/1hO2PEFz2Pw-DU8xNkGEBr-jYHSmOZbiX?usp=sharing

## Expected behavior

The notebook should execute much faster when executing `torch.tensor.item()`. 

## Environment

 - Reproducible on XLA backend [CPU/TPU]: TPU
 - torch_xla version: nightly
 - Any other relevant information. In case of issues related to slow executions, re-compilations, etc.., it is useful to have the `tar.gz` report from https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information

## Additional context

**IMPORTANT**: 
- I currently have to set the variable `os.environ["XLA_USE_32BIT_LONG"] = "1"` to avoid the error:

```
Unimplemented: From /job:tpu_worker/replica:0/task:0:
While rewriting computation to not contain X64 element types, XLA encountered an HLO for which this rewriting is not implemented: %sort.15994 = (s64[1,2,128]{2,1,0}, s32[1,2,128]{2,1,0}) sort(s64[1,2,128]{2,1,0} %add.15983, s32[1,2,128]{2,1,0} %iota.15984), dimensions={2}, to_apply=%compare-less-than.15985
	 [[{{node XRTCompile}}]]
```

- The Reformer model is quite a complicated model that was recently added to `transformers`: https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_reformer.py . It calls `torch.gather()` and `torch.sort()` multiple times in the forward pass, calls `torch.manual_seed()` at every forward pass and uses a customized backward pass.

I would be very thankful for any tips!

**Comment by patrickvonplaten** [comment]
Also here is the metrics report for the code as  shown above:

```
Time: 0.3164668083190918
Metric: IrValueTensorToXlaData
  TotalSamples: 82
  Accumulator: 252ms985.338us
  ValueRate: 501ms255.151us / second
  Rate: 163.116 / second
  Percentiles: 1%=897.252us; 5%=001ms017.831us; 10%=001ms114.246us; 20%=001ms222.341us; 50%=002ms397.334us; 80%=005ms218.487us; 90%=006ms394.029us; 95%=007ms008.324us; 99%=013ms599.874us
Metric: OutboundData
  TotalSamples: 98
  Accumulator: 9.62MB
  ValueRate: 16.94MB / second
  Rate: 172.535 / second
  Percentiles: 1%=4.00B; 5%=4.00B; 10%=4.00B; 20%=512.00B; 50%=1.00KB; 80%=128.00KB; 90%=512.00KB; 95%=512.00KB; 99%=640.00KB
Metric: TransferToServerTime
  TotalSamples: 98
  Accumulator: 282ms508.726us
  ValueRate: 503ms681.058us / second
  Rate: 174.995 / second
  Percentiles: 1%=893.451us; 5%=982.027us; 10%=001ms056.636us; 20%=001ms186.419us; 50%=002ms066.217us; 80%=004ms079.310us; 90%=006ms361.707us; 95%=007ms003.441us; 99%=013ms533.078us
Metric: TransferToServerTransformTime
  TotalSamples: 98
  Accumulator: 018ms345.054us
  ValueRate: 032ms296.848us / second
  Rate: 172.531 / second
  Percentiles: 1%=077.077us; 5%=097.808us; 10%=103.941us; 20%=111.536us; 50%=135.941us; 80%=209.933us; 90%=253.811us; 95%=364.391us; 99%=002ms490.003us
Counter: CreateDataHandles
  Value: 98
Counter: CreateXlaTensor
  Value: 13494
Counter: DestroyXlaTensor
  Value: 13289
Counter: DeviceDataCacheMiss
  Value: 15
Counter: XRTAllocateFromTensor_Empty
  Value: 14
Counter: XrtSessionCount
  Value: 1
Counter: xla::_log_softmax
  Value: 5
Counter: xla::_log_softmax_backward_data
  Value: 5
Counter: xla::_s_where
  Value: 45
Counter: xla::_unsafe_view
  Value: 455
Counter: xla::add
  Value: 446
Counter: xla::add_
  Value: 433
Counter: xla::addcmul
  Value: 125
Counter: xla::arange_out
  Value: 110
Counter: xla::argmax
  Value: 15
Counter: xla::as_strided
  Value: 160
Counter: xla::bernoulli_
  Value: 225
Counter: xla::bmm
  Value: 255
Counter: xla::cat
  Value: 395
Counter: xla::copy_
  Value: 530
Counter: xla::div
  Value: 61
Counter: xla::div_
  Value: 225
Counter: xla::embedding
  Value: 5
Counter: xla::embedding_dense_backward
  Value: 5
Counter: xla::empty
  Value: 975
Counter: xla::empty_strided
  Value: 160
Counter: xla::exp
  Value: 90
Counter: xla::expand
  Value: 540
Counter: xla::fill_
  Value: 5
Counter: xla::gather
  Value: 150
Counter: xla::index_select
  Value: 5
Counter: xla::logsumexp
  Value: 60
Counter: xla::mean
  Value: 30
Counter: xla::mm
  Value: 675
Counter: xla::mul
  Value: 840
Counter: xla::native_batch_norm
  Value: 125
Counter: xla::native_batch_norm_backward
  Value: 65
Counter: xla::native_layer_norm
  Value: 125
Counter: xla::native_layer_norm_backward
  Value: 65
Counter: xla::ne
  Value: 30
Counter: xla::neg
  Value: 45
Counter: xla::nll_loss_backward
  Value: 5
Counter: xla::nll_loss_forward
  Value: 5
Counter: xla::normal_
  Value: 15
Counter: xla::permute
  Value: 165
Counter: xla::pow
  Value: 60
Counter: xla::relu
  Value: 60
Counter: xla::remainder
  Value: 60
Counter: xla::repeat
  Value: 90
Counter: xla::rsqrt
  Value: 60
Counter: xla::scatter_
  Value: 30
Counter: xla::scatter_add_
  Value: 30
Counter: xla::slice
  Value: 1725
Counter: xla::sort
  Value: 30
Counter: xla::split
  Value: 10
Counter: xla::sqrt
  Value: 30
Counter: xla::squeeze
  Value: 30
Counter: xla::sub
  Value: 215
Counter: xla::sum
  Value: 250
Counter: xla::t
  Value: 845
Counter: xla::threshold_backward
  Value: 30
Counter: xla::transpose
  Value: 455
Counter: xla::unsqueeze
  Value: 245
Counter: xla::view
  Value: 3420
Counter: xla::zero_
  Value: 430
Metric: XrtAllocateFromTensor
  TotalSamples: 98
  Accumulator: 041ms505.612us
  Mean: 413.323us
  StdDev: 189.212us
  Rate: 175.019 / second
  Percentiles: 25%=230.896us; 50%=377.206us; 80%=606.899us; 90%=679.489us; 95%=737.291us; 99%=952.535us
```

**Comment by patrickvonplaten** [comment]
This issues seems very much related: https://github.com/pytorch/xla/issues/1775

**Comment by dlibenzi** [comment]
Calling `item()` forces extra graph executions, which does indeed slow down performance.
This is why we suggest (unless one does not care of execution time) to do that once every ~20 steps:

https://github.com/pytorch/xla/blob/3b4b5790ba9e9141f19b0e33b3a7a7cd1bad8ead/test/test_train_mp_mnist.py#L129

But the code above is not correct, as it should use `xm.optimizer_step(optimizer)`.

https://github.com/pytorch/xla/blob/3b4b5790ba9e9141f19b0e33b3a7a7cd1bad8ead/test/test_train_mp_mnist.py#L127

But in general, IIRC, Reformer, at least in the state we saw it a few months ago, was not a friendly model for TPUs as it has the inner sorting loop generating a lot of dynamic shapes and triggering constant recompilations.
In the notebook I fished out of my Trash can I noticed I also had those set:
```Python
os.environ['XLA_USE_32BIT_LONG'] = '1'
os.environ['XLA_USE_BF16'] = '1'
os.environ['TRIM_GRAPH_SIZE'] = '1000000'
```

I am not sure whether the code changed since then, and whether it is possible to tweak the model to generate a constant set of shapes (bucketing).

**Comment by patrickvonplaten** [comment]
Hi @dlibenzi, 

Thanks a lot for answering this quickly! I also think it is because Reformer is not TPU friendly since other models compile very quickly using the exact same code. 

What exactly do you mean by "generate a constant set of shapes (bucketing)" ? 

I think what's special in Reformer is:

1. The tensors are reshaped a lot (lots of `torch.perm()`, `torch.reshape()`, `torch.flatten()`)
2. `torch.sort()` is used
3. `torch.gather()` is used multiple times

From the points 1-3, are any of these things known to hurt performane a lot?

**Comment by dlibenzi** [comment]
So I create a notebook with slight modification to yours, and with constant input (like your example), compilation stabilized and then runs fast:

https://colab.research.google.com/drive/1kMAT2Xgs-J2rrnvGt5xfi-ZA43QtZxFd?usp=sharing

Maybe there is some hope is bucketing inputs.

Bucketing, in the NLP world, mean choosing a small set (3..4) of max-sequence-length buckets, and the fitting the samples into the proper ones by padding.

Adding @jysohn23 which is the engineer following HF model optimizations for TPU.

**Comment by patrickvonplaten** [comment]
Thanks for your help @dlibenzi , I works well for shorter sequence lengths but does not seem to stabilize for longer sequence lengths. I will take a more in-detail look into the hugging face reformer code. 

Here a notebook that shows that for longer sequence length the code is quite slow:
https://colab.research.google.com/drive/1jZcmWqOaYCHRwl0yBqc_8fT4UtPpSurf
